src/zcap.rs

```diff
@@ -0,0 +1,473 @@
+use crate::one_or_many::OneOrMany;
+use crate::rdf::DataSet;
+use crate::vc::{Check, LinkedDataProofOptions, Proof, ProofPurpose, VerificationResult, URI};
+
+use async_trait::async_trait;
+use chrono::prelude::*;
+use serde::{Deserialize, Serialize};
+use serde_json::Value;
+
+const DEFAULT_CONTEXT: &str = SECURITY_V2_CONTEXT;
+
+// limited initial definition of a ZCAP Delegation, generic over Action and Caveat types
+#[derive(Debug, Serialize, Deserialize, Clone, Default)]
+#[serde(rename_all = "camelCase")]
+pub struct Delegation<A, C> {
+    #[serde(rename = "@context")]
+    pub context: Contexts,
+    pub id: URI,
+    pub parent_capability: URI,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub invoker: Option<URI>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub capability_action: Option<A>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub caveat: Option<C>,
+    // This field is populated only when using
+    // embedded proofs such as LD-PROOF
+    //   https://w3c-ccg.github.io/ld-proofs/
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub proof: Option<Proof>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    #[serde(flatten)]
+    pub property_set: Option<Map<String, Value>>,
+}
+
+impl<A, C> Delegation<A, C>
+where
+    A: Serialize + Send + Sync + Clone,
+    C: Serialize + Send + Sync + Clone,
+{
+    pub async fn verify(
+        &self,
+        _options: Option<LinkedDataProofOptions>,
+        resolver: &dyn DIDResolver,
+    ) -> VerificationResult {
+        match &self.proof {
+            None => VerificationResult::error("No applicable proof"),
+            Some(proof) => {
+                let mut result = proof.verify(self, resolver).await;
+                if result.errors.is_empty() {
+                    result.checks.push(Check::Proof);
+                }
+                result
+            }
+        }
+    }
+
+    // https://w3c-ccg.github.io/ld-proofs/
+    pub async fn generate_proof(
+        &self,
+        jwk: &JWK,
+        options: &LinkedDataProofOptions,
+    ) -> Result<Proof, Error> {
+        LinkedDataProofs::sign(self, options, jwk).await
+    }
+
+    /// Prepare to generate a linked data proof. Returns the signing input for the caller to sign
+    /// and then pass to [`ProofPreparation::complete`] to complete the proof.
+    pub async fn prepare_proof(
+        &self,
+        public_key: &JWK,
+        options: &LinkedDataProofOptions,
+    ) -> Result<ProofPreparation, Error> {
+        LinkedDataProofs::prepare(self, options, public_key).await
+    }
+
+    pub fn set_proof(self, proof: Proof) -> Self {
+        Self {
+            proof: Some(proof),
+            ..self
+        }
+    }
+}
+
+#[async_trait]
+impl<A, C> LinkedDataDocument for Delegation<A, C>
+where
+    A: Serialize + Send + Sync + Clone,
+    C: Serialize + Send + Sync + Clone,
```

Comment:
There is a CI failure having to do with Send in wasm:
https://github.com/spruceid/ssi/pull/189/checks#step:12:187
```
error[E0053]: method `to_dataset_for_signing` has an incompatible type for trait
   --> src/zcap.rs:103:5
    |
103 | /     async fn to_dataset_for_signing(
104 | |         &self,
105 | |         parent: Option<&(dyn LinkedDataDocument + Sync)>,
106 | |     ) -> Result<DataSet, Error> {
...   |
115 | |         json_to_dataset(&json, more_contexts.as_ref(), false, None, &mut loader).await
116 | |     }
    | |_____^ expected trait `futures::Future<Output = Result<DataSet, error::Error>>`, found trait `futures::Future<Output = Result<DataSet, error::Error>> + std::marker::Send`
```
Could something similar to what was done with `async_trait` be needed here? We are using this in places:
```
#[cfg_attr(target_arch = "wasm32", async_trait(?Send))]
#[cfg_attr(not(target_arch = "wasm32"), async_trait)]
```
I'm not sure what is the best way to abstract this, but I think the Send trait bound has to be removed for the wasm32 target.

Reply:
was missing a wasm_bindgen attribute